DOCSP-40794: Bulk write guide (#37)
mongodb/docs-cpp: source/write/bulk-write.txt

Define the Write Operations
---------------------------

For each write operation you want to perform, create an instance of one of
the following model classes:

- ``mongocxx::model::insert_one``
- ``mongocxx::model::update_one``
- ``mongocxx::model::update_many``
- ``mongocxx::model::replace_one``
- ``mongocxx::model::delete_one``
- ``mongocxx::model::delete_many``

Then, append each write model to the ``mongocxx::bulk_write`` instance returned
by the ``create_bulk_write()`` method.

The following sections show how to create and use instances of the preceding
write model classes.

.. _cpp-bulk-insert-model:

Insert Operations
~~~~~~~~~~~~~~~~~

To perform an insert operation, create an instance of the ``mongocxx::model::insert_one``
class and specify the document you want to insert. Then, append the model instance
to an instance of the ``mongocxx::bulk_write`` class.

The following example creates an instance of ``mongocxx::model::insert_one`` and appends
it to a ``mongocxx::bulk_write`` instance called ``bulk``:

.. literalinclude:: /includes/write/bulk-write.cpp
   :start-after: start-bulk-insert-one
   :end-before: end-bulk-insert-one
   :language: cpp
   :dedent:

> To insert multiple documents, create an instance of ``mongocxx::model::insert_many``

Reviewer: ```suggestion
To insert multiple documents, create an instance of ``mongocxx::model::insert_one``
```

Author: good catch!!

for each document.

.. _cpp-bulk-update-model:

Update Operations
~~~~~~~~~~~~~~~~~

To update a document, create an instance of ``mongocxx::model::update_one`` and
pass in the following arguments:

> - **Query filter** document, which specifies the criteria used to match documents

Reviewer: ```suggestion
-  A **query filter** document, which specifies the criteria used to match documents
```

Author: It seems a little unnatural, but we [don't start bulleted lists with articles](https://www.mongodb.com/docs/meta/style-guide/style/lists/#list-items)

  in your collection.

> - **Update** document, which specifies the kind of update to perform. For more information

Reviewer: ```suggestion
-  An **update** document, which specifies the kind of update to perform. For more information
```

  about update operations, see the :manual:`Field Update Operators
  </reference/operator/update-field/>` guide in the {+mdb-server+} manual.

> This model instructs the driver to update *the first* document that matches your

Reviewer: Q: Why is this information in the middle of the steps of which arguments to pass in/how to append the model? 

S: Move this information to the beginning of the section about update_one, such as:

To update a single document, create an instance of ``mongocxx::model::update_one`. This model instructs the driver...

and pass in the same arguments. This model instructs the driver to update *all* documents
that match your query filter.

The following example creates an instance of ``mongocxx::model::update_many`` and appends
it to ``bulk``:

.. literalinclude:: /includes/write/bulk-write.cpp
   :start-after: start-bulk-update-many
   :end-before: end-bulk-update-many
   :language: cpp
   :dedent:

.. _cpp-bulk-replace-model:

Replace Operations
~~~~~~~~~~~~~~~~~~

A replace operation removes all fields and values of a specified document and
replaces them with new ones. To perform a replace operation, create an instance
of the ``mongocxx::model::replace_one`` class and pass it a query filter and
the fields and values you want to store in the matching document. Then, append
the model instance to an instance of the ``mongocxx::bulk_write`` class.

The following example creates an instance of ``mongocxx::model::replace_one`` and appends
it to a ``mongocxx::bulk_write`` instance called ``bulk``:

.. literalinclude:: /includes/write/bulk-write.cpp
   :start-after: start-bulk-replace-one
   :end-before: end-bulk-replace-one
   :language: cpp
   :dedent:

> To replace multiple documents, you must create an instance of ``mongocxx::model::replace_one``

Reviewer: ```suggestion
To replace multiple documents, you must create a new instance of ``mongocxx::model::replace_one``
```

   :start-after: start-bulk-delete-many
   :end-before: end-bulk-delete-many
   :language: cpp
   :dedent:

Run the Bulk Operation
----------------------

To run a bulk operation, call the ``execute()`` method on an instance of the 
``mongocxx::bulk_write`` class that contains your write models. By default, the
``execute()`` method runs the operations in the order they're appended to the
``mongocxx::bulk_write`` instance.

The following example performs the :ref:`insert <cpp-bulk-insert-model>`,
:ref:`update <cpp-bulk-update-model>`, :ref:`replace <cpp-bulk-replace-model>`,
and :ref:`delete <cpp-bulk-delete-model>` operations specified in the preceding
sections of this guide by appending each corresponding write model to an instance
of ``mongocxx::bulk_write`` and calling the ``execute()`` method. Then, it prints
the number of modified documents:

.. io-code-block::

   .. input:: /includes/write/bulk-write.cpp
      :start-after: start-bulk-run
      :end-before: end-bulk-run
      :language: cpp
      :dedent:
      :emphasize-lines: 30

   .. output::
      
      Modified documents: 2

If any of the write operations fail, the {+driver-short+} raises a

> ``mongocxx::bulk_write_exception`` and does not perform any further operations.

Reviewer: Q: Does the driver still successfully perform the operations that were ordered before the failing write operation? Or do all operations fail if just one fails/if an exception is thrown?

Author: Yes, the operations before the failing one still apply

an instance of the ``mongocxx::options::bulk_write`` class as a parameter. The following
table describes the fields you can set in a ``mongocxx::options::bulk_write``
instance:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Field
     - Description

   * - ``ordered``
     - | If ``true``, the driver performs the write operations in the order
         provided. If an error occurs, the remaining operations are not
         attempted. 
       |
       | If ``false``, the driver performs the operations in an
         arbitrary order and attempts to perform all operations.
       | Defaults to ``true``.

   * - ``bypass_document_validation``
     - | Specifies whether the operation bypasses document-level validation. For more
         information, see :manual:`Schema
         Validation </core/schema-validation/#schema-validation>` in the MongoDB
         Server manual.
       | Defaults to ``false``.

   * - ``write_concern``
     - | Specifies the write concern for the bulk operation. For more information, see
         :manual:`Write Concern </reference/write-concern/>` in the {+mdb-server+} manual.

   * - ``comment``

>      - | A comment to attach to the operation. For more information, see the :manual:`delete command

Reviewer: S: Mirror the other entries in the table by using a verb? Ex: Attaches a comment...

         fields </reference/command/delete/#command-fields>` guide in the
         {+mdb-server+} manual.

   * - ``let``

>      - | A map of parameter names and values. Values must be constant or closed

Reviewer: Same comment as above, start with a verb

.. note::

   Unordered bulk operations do not guarantee order of execution. The order can
   differ from the way you list them to optimize the runtime.

.. _cpp-bulk-return-value:

Return Value
------------

The ``execute()`` method returns an instance of the ``mongocxx::result::bulk_write`` class.
The ``mongocxx::result::bulk_write`` class contains the following member functions:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Function
     - Description
         
   * - ``deleted_count``
     - | The number of documents deleted, if any.

   * - ``inserted_count``
     - | The number of documents inserted, if any.

   * - ``matched_count``
     - | The number of documents matched for an update, if applicable.

   * - ``modified_count``
     - | The number of documents modified, if any.

   * - ``upserted_count``
     - | The number of documents upserted, if any.

   * - ``upserted_ids``
     - | A map of the operation's index to the ``_id`` of the upserted documents, if
         applicable.

Additional Information
----------------------

To learn how to perform individual write operations, see the following guides:

- :ref:`cpp-write-insert`

> - :ref:`cpp-write-update`

Author: Note to reviewer: this link will be broken until #34 is merged
